hela/model.py

```diff
@@ -0,0 +1,296 @@
+                                             names=(self.dataset.names +
+                                                    ["joint prediction"]),
+                                             colors=(self.dataset.colors +
+                                                     ["C0"]))
+
+        fig.savefig(os.path.join(self.output_path, "feature_importances.pdf"),
+                    bbox_inches='tight')
+        return fig, axes
+
+    def predict(self, quiet=False):
+        """
+        Predict values from the trained random forest.
+
+        Parameters
+        ----------
+        plot_posterior : bool
+
+        Returns
+        -------
+        preds : `~numpy.ndarray`
+            ``N x M`` values where ``N`` is number of parameters, ``M`` is
+            number of samples/trees (check out attributes of model for
+            metadata)
+        """
+        if self._posterior is None:
+            model_file = os.path.join(self.model_path, "model.pkl")
+            # Loading random forest from '{}'...".format(model_file)
+            model = joblib.load(model_file)
+
+            # Loading data from '{}'...".format(data_file)
+            data, _ = load_data_file(self.data_file, model.rf.n_features_)
+
+            posterior = model.posterior(data[0])
+
+            if not quiet:
+                posterior_ranges = data_ranges(posterior)
+                for name_i, pred_range_i in zip(model.names, posterior_ranges):
+                    print("Prediction for {}: {:.3g} "
+                          "[+{:.3g} -{:.3g}]".format(name_i, *pred_range_i))
+
+            self._posterior = posterior
+
+        return self._posterior
+
+    def plot_posterior(self):
+        """
+        Plot the posterior distributions for each parameter.
+
+        Returns
+        -------
+        fig, axes
+        """
+        model_file = os.path.join(self.model_path, "model.pkl")
+        # Loading random forest from '{}'...".format(model_file)
+        model = joblib.load(model_file)
+
+        fig, axes = plot_posterior_matrix(self._posterior,
+                                          names=model.names,
+                                          ranges=model.ranges,
+                                          colors=model.colors)
+        os.makedirs(self.output_path, exist_ok=True)
+        fig.savefig(os.path.join(self.output_path, "posterior_matrix.pdf"),
+                    bbox_inches='tight')
+        return fig, axes
+
+
+def data_ranges(posterior, percentiles=(50, 16, 84)):
+    """
+    Return posterior ranges.
+
+    Parameters
+    ----------
+    posterior : `~numpy.ndarray`
+    percentiles : tuple
+
+    Returns
+    -------
+    ranges : `~numpy.ndarray`
+    """
+    values = wpercentile(posterior.samples, posterior.weights,
+                         percentiles, axis=0)
+    ranges = np.array(
+        [values[0], values[2] - values[0], values[0] - values[1]])
+    return ranges.T
+
+
+def generate_example_data():
```

Comment:
I've been trying to craft an example with correlations between slopes and ints but haven't been able to create a good working example. Would you be able to give me some more hints how to demonstrate this?

Reply:
Hi @bmorris3,

Sorry, I have realized that my comment was very vague and unclear. Something like this might work:

```python
def generate_example_data():
    """
    Generate an example dataset in the new directory ``linear_dataset``.

    Returns
    -------
    example_dir : str
        Path to the directory of the example data
    training_dataset : str
        Path to the dataset metadata JSON file
    samples : `~numpy.ndarray`
    """
    example_dir = 'linear_dataset'
    training_dataset = os.path.join(example_dir, 'example_dataset.json')

    os.makedirs(example_dir, exist_ok=True)

    # Save the dataset metadata to a JSON file
    dataset = {
        "metadata": {
            "names": ["slope", "intercept"],
            "ranges": [[0, 1], [0, 1]],
            "colors": ["#F14532", "#4a98c9"],
            "num_features": 1000
        },
        "training_data": "training.npy",
        "testing_data": "testing.npy"
    }

    with open(training_dataset, 'w') as fp:
        json.dump(dataset, fp)

    # Generate fake training data
    npoints = 20000

    slopes = np.random.uniform(size=npoints)
    ints = np.random.uniform(size=npoints)
    x = np.linspace(0, 1, 1000)[:, np.newaxis]

    # Add correlated noise to parameters to introduce degeneracies
    noise_ints = np.random.normal(scale=0.15, size=npoints)
    noise_slopes = np.abs(noise_ints) + np.random.normal(scale=0.02, size=npoints)

    # Add also noise to data points (not strictly necessary)
    data = ((slopes + noise_slopes) * x + (ints + noise_ints) +
            np.random.normal(scale=0.01, size=(1000, npoints)))

    labels = np.vstack([slopes, ints])
    X = np.vstack([data, labels])

    # Split dataset into training and testing segments
    training = X[:, :int(0.8 * npoints)].T
    testing = X[:, int(0.8 * npoints):].T

    np.save(os.path.join(example_dir, 'training.npy'), training)
    np.save(os.path.join(example_dir, 'testing.npy'), testing)

    # Generate a bunch of samples with a test value to "retrieve" with the
    # random forest:
    true_slope = 0.7
    true_intercept = 0.5

    samples = true_slope * x + true_intercept
    return training_dataset, example_dir, samples.T[0]
```

The posteriors would look like this:

![version3](https://user-images.githubusercontent.com/1254337/68867383-03946600-06f6-11ea-9191-f14898bdf214.png)